### PR TITLE
[serve] Add serve.get_deployment() API

### LINKED ISSUE #14955: Make it easier to update options on running deployments from a different script
e.g., want to just scale the number of replicas

### PR DESCRIPTION
**Pull request by edoakes**
## Why are these changes needed?

Makes it easier to update deployments from a new script. For example, to scale replicas:
```python
serve.get_deployment(name).options(num_replicas=10).deploy()
```

## Related issue number

Closes https://github.com/ray-project/ray/issues/14955

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by edoakes** [comment]
FYI this is rebased off of https://github.com/ray-project/ray/pull/14935

**Comment by edoakes** [comment]
This is ready for review.

File: python/ray/serve/api.py [hunk 1]
```diff
@@ -1306,3 +1310,27 @@ def decorator(backend_def):
...
 
     return decorator
+
+
+def get_deployment(name: str):
+    """Retrieve RayServeHandle for service endpoint to invoke it from Python.
+
+    Args:
+        name(str): name of the deployment. This must have already been.
+        deployed
```
Comment (architkulkarni) [inline python/ray/serve/api.py:1320]:
```suggestion
    Args:
        name(str): name of the deployment. This must have already been
        deployed.
```

File: python/ray/serve/api.py [hunk 2]
```diff
@@ -1306,3 +1310,27 @@ def decorator(backend_def):
             ray_actor_options=ray_actor_options)
 
     return decorator
+
+
+def get_deployment(name: str):
```
Comment (architkulkarni) [inline python/ray/serve/api.py:1315]: Can we add a return type annotation here?

File: python/ray/serve/backend_state.py [hunk 3]
```diff
@@ -539,57 +539,50 @@ def get_backend(self, backend_tag: BackendTag) -> Optional[BackendInfo]:
...
         # Ensures this method is idempotent.
-        backend_info = self._backend_metadata.get(backend_tag)
-        if backend_info is not None:
+        existing_info = self._backend_metadata.get(backend_tag)
+        if existing_info is not None:
             # Old codepath.
-            if version is None:
-                if (backend_info.backend_config == backend_config
-                        and backend_info.replica_config == replica_config):
+            if backend_info.version == RESERVED_VERSION_TAG:
```
Comment (architkulkarni) [inline python/ray/serve/backend_state.py:569]: Could you please add a comment somewhere explaining RESERVED_VERSION_TAG?

File: python/ray/serve/api.py [hunk 4]
```diff
@@ -513,6 +513,10 @@ def deploy(self,
     def delete_deployment(self, name: str) -> None:
         self._wait_for_goal(self._controller.delete_deployment.remote(name))
 
+    @_ensure_connected
+    def get_deployment_info(self, name: str) -> Tuple[BackendInfo, str, str]:
```
Comment (architkulkarni) [inline python/ray/serve/api.py:517]:
```suggestion
    def get_deployment_info(self, name: str) -> Tuple[BackendInfo, str]:
```

File: python/ray/serve/endpoint_state.py [hunk 5]
```diff
@@ -108,6 +108,12 @@ def shadow_traffic(self, endpoint: EndpointTag, backend: BackendTag,
         self._checkpoint()
         self._notify_traffic_policies_changed(endpoint)
 
+    def get_endpoint_route(self, endpoint: EndpointTag) -> Optional[str]:
```
Comment (architkulkarni) [inline python/ray/serve/endpoint_state.py:111]: Do we need to worry about variable routes here?  If there's some nontrivial interaction, maybe we should document it and add a test.

**Review by architkulkarni — CHANGES_REQUESTED** [review]
Looks good!  Just had some minor comments and questions

File: python/ray/serve/endpoint_state.py (hunk 5 above)
Comment (architkulkarni) [reply python/ray/serve/endpoint_state.py:111]: Actually I think this is fine, we're just matching the endpoint tag, not the route itself

File: python/ray/serve/tests/test_get_deployment.py [hunk 6]
```diff
@@ -0,0 +1,141 @@
+import os
```
Comment (simon-mo) [inline python/ray/serve/tests/test_get_deployment.py:1]: any reason not to just include this in test_deploy?

**Review by simon-mo — APPROVED** [review]

File: python/ray/serve/tests/test_get_deployment.py (hunk 6 above)
Comment (edoakes) [reply python/ray/serve/tests/test_get_deployment.py:1]: I like the logical separation but I just discovered that you can group tests in a class; let me do that instead

**Comment by edoakes** [comment]
@architkulkarni your comments are addressed

**Review by architkulkarni — APPROVED** [review]

**Pull request merged**